src/lib/optimizer/strategy/predicate_merge_rule.cpp

```diff
@@ -0,0 +1,127 @@
+      } else {
+        // New node for a known subplan found
+        subplan_roots.insert(std::make_pair(node, subplan_roots[*parent]));
+      }
+      if (node->type == LQPNodeType::Union) {
+        union_node_counts[subplan_roots[node]]++;
+      }
+    }
+    return LQPVisitation::VisitInputs;
+  });
+
+  for (const auto& node_and_count : union_node_counts) {
+    if (node_and_count.second >= optimization_threshold) {
+      _merge_subplan(node_and_count.first, std::nullopt);
+    }
+  }
+}
+
+/**
+ * Merge a subplan that only consists of PredicateNodes and UnionNodes into a single PredicateNode. The
+ * subsequent_expression parameter passes the translated expressions to the translation of its children nodes, which
+ * allows to add the translated expression of child node before its parent node to the output expression.
+ *
+ * A subplan consists of linear "chain" and forked "diamond" parts.
+ *
+ * EXAMPLE:
+ *         Step 1                   Step 2                   Step 3                         Step 4
+ *
+ *           |                        |                        |                              |
+ *      ___Union___              ___Union___           Predicate (A OR B)      Predicate ((D AND C) AND (A OR B))
+ *    /            \            /           \                  |                              |
+ * Predicate (A)   |         Predicate (A)  |                  |
+ *    |            |           |            |                  |
+ *    |       Predicate (B)    |      Predicate (B)            |
+ *    \           /            \          /                    |
+ *     Predicate (C)          Predicate (D AND C)     Predicate (D AND C)
+ *           |                        |                        |
+ *     Predicate (D)
+ *           |
+ */
+std::shared_ptr<AbstractExpression> PredicateMergeRule::_merge_subplan(
+    const std::shared_ptr<AbstractLQPNode>& begin,
+    const std::optional<const std::shared_ptr<AbstractExpression>>& subsequent_expression) const {
+  switch (begin->type) {
+    case LQPNodeType::Predicate: {
+      const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(begin);
+      auto expression = predicate_node->predicate();
+      if (subsequent_expression && begin->output_count() == 1) {
+        // Only merge inside a chain
+        expression = and_(expression, *subsequent_expression);
+      }
+
+      const auto left_input_expression = _merge_subplan(begin->left_input(), expression);
+      if (begin->left_input()->output_count() == 1 && left_input_expression) {
+        // Only merge inside a chain. Nodes at the bottom of a diamond would unnecessarily inflate the resulting logical
+        // expression now.
+        lqp_remove_node(begin->left_input());
+        predicate_node->node_expressions[0] = left_input_expression;
+        return left_input_expression;
+      } else {
+        return expression;
+      }
+    }
+
+    case LQPNodeType::Union: {
```

Comment:
This is the case with which everything starts, correct? I'd place it first for better readability.